File: interviews/lc/03-rotate-array.py
from typing import List


def reverse(nums: List[int], leftIdx: int, rightIdx: int) -> None:
    while leftIdx < rightIdx:
        nums[leftIdx], nums[rightIdx] = nums[rightIdx], nums[leftIdx]
        leftIdx += 1
        rightIdx -= 1


def rotate(nums: List[int], k: int) -> None:
    k = k % len(nums)
    if k > 0:
        reverse(nums, 0, len(nums) - 1)
        reverse(nums, 0, k - 1)
        reverse(nums, k, len(nums) - 1)


def runTestCase(nums: List[int], k: int) -> None:
    origList = ",".join([str(n) for n in nums])
    rotate(nums, k)
    rotatedList = ",".join([str(n) for n in nums])
    print(f"{origList}, k={k} => {rotatedList}")


if __name__ == "__main__":
    runTestCase([1], 0)
    runTestCase([1], 10)
    runTestCase([1, 2], 1)
    runTestCase([1, 2], 2)
    runTestCase([1, 2], 3)
    runTestCase([1, 2, 3, 4, 5, 6, 7], 3)

File: interviews/lc/03-rotate-array.cpp
#include <iostream>
#include <string>
#include <vector>
#include "lc-utility.h"

using std::string;
using std::vector;

void reverse(vector<int>& nums, vector<int>::iterator left,
             vector<int>::iterator right) {
  while (left < right) {
    const auto tmp = *left;
    *left = *right;
    *right = tmp;
    left++;
    right--;
  }
}

void rotate(vector<int>& nums, size_t k) {
  k = k % nums.size();
  if (k > 0) {
    reverse(nums, nums.begin(), nums.end() - 1);
    reverse(nums, nums.begin(), nums.begin() + k - 1);
    reverse(nums, nums.begin() + k, nums.end() - 1);
  }
}

void runTestCase(vector<int> nums, size_t k) {
  const auto originalArr = vector2str(nums);
  rotate(nums, k);
  const auto rotatedArr = vector2str(nums);
  std::cout << originalArr << " => " << rotatedArr << " k=" << k << std::endl;
}

int main(int argc, char* argv[]) {
  runTestCase({1}, 0);
  runTestCase({1}, 10);
  runTestCase({1, 2}, 1);
  runTestCase({1, 2}, 2);
  runTestCase({1, 2}, 3);
  runTestCase({1, 2, 3, 4, 5, 6, 7}, 3);
  return EXIT_SUCCESS;
}